DP-1657 Add update metadata function to base image

MatMoore opened the pull request:
This PR is preparation for adding an "update metadata" endpoint.

I've added some functionality to the metadata class to detect whether an update is minor or not; and if it is minor, to create a new minor version (we'll deal with major updates later).

This is implemented at the DataProductMetadata level rather than in the base class, since the schema update case is more complicated and until we start working on that, I'm not sure what we'll be able to reuse from this implementation.

I've also moved the `format_error_response` function into the base since I saw we were creating this in a few different places. For consistency, I made format_response_json also do the `json.dumps` which is a breaking change but very easy to update.

MatMoore committed:
Move error function

Move format_error_response into data_platform_api_responses from
create_schema lambda.

MatMoore committed:
Document the existing behaviour of BaseJsonSchema

Also ensure latest_version_saved_data is always initialised, even if
it's none.

MatMoore committed:
Add update functionality to DataProductMetadata

MatMoore committed:
Ensure environment variables are set before code is imported

murdo-moj commented on containers/daap-python-base/src/var/task/data_product_metadata.py:
```diff
@@ -332,23 +418,71 @@ class to handle creation and updating of
     """
 
-    # returns path of latest metadata or v1 if it doesn't exist
+    MINOR_UPDATE_FIELDS = {
```
Changing any data product level (non schema level) fields which are editable by the user will trigger a minor version update I think. Looking through the spec, only schema changes would be a breaking change for the user.

MatMoore replied:
That makes sense, although I'm assuming some of these are not updatable at all (e.g. name)

https://github.com/ministryofjustice/modernisation-platform-environments/blob/main/terraform/environments/data-platform/data-product-metadata-json-schema/v1.1.0/moj_data_product_metadata_spec.json

The full list is

name
description
domain
dataProductOwner
dataProductOwnerDisplayName
email
status
dpiaRequired
retentionPeriod
dataProductMaintainer
dataProductMaintainerDisplayName
tags
version
id
lastUpdatedDate
creationDate
s3Location
rowCount
schema

So I think I need to add in:
domain
status
dpiaRequired
retentionPeriod
dataProductMaintainer
dataProductMaintainerDisplayName
tags

We also need to make sure that any generated fields are preserved when updating, since they won't be passed into the API call. But since none of that is implemented I think we can skip that for now.

MatMoore replied:
Just pushed a new commit with some changes to the fields

LavMatt commented on containers/daap-python-base/src/var/task/data_product_metadata.py:
```diff
@@ -332,23 +418,71 @@ class to handle creation and updating of
+                version=new_version,
+            )
+            new_metadata.write_json_to_s3()
+            return new_metadata
```
might be personal preference but i kind of liked enforcing the write method to be passed a path, and think i'd find something like below clearer
```suggestion
            new_version_key = (
                DataProductConfig(self.data_product_name).metadata_path(new_version).key
            )
            self.write_json_to_s3(new_version_key)
            return self.data

MatMoore replied:
Fair enough, I'll just drop that commit then, as this is what it looked like before.

It makes me wonder though if we should lose the `bucket_path` parameter altogether, so that `load` is consistent with `write_json_to_s3` (both are called with a path)?

LavMatt replied:
as in the get rid of `latest_version_key`? 

I think `load` could maybe be more suitably named, like `load_latest_version` but if there might be any use for loading in older versions then possibly but i can't think of any at the moment, otherwise we'd always want to load the `latest_version_key`

LavMatt reviewed (commented):
This looks good and sets a nice framework i can work into the schema updates (which i'm looking at now), but couple of points:
1. is there a reason we need to distinguish, once we have the update type, whether it's minor or major to write it? For the metadata (and table schema) all that really differs is the path, although with major we would need to follow that up-versioning to the data as well i guess (so maybe that's why actually).
2. Even with minor updates we'll still need to write all associated schema for the data product into the new version folder which isn't currently happening. I wrote this function to do that https://github.com/ministryofjustice/data-platform/blob/main/containers/daap-create-schema/src/var/task/create_schema.py#L18 for schema

MatMoore committed:
Add more updatable fields

MatMoore commented:
> This looks good and sets a nice framework i can work into the schema updates (which i'm looking at now), but couple of points:
> 
> 1. is there a reason we need to distinguish, once we have the update type, whether it's minor or major to write it? For the metadata (and table schema) all that really differs is the path, although with major we would need to follow that up-versioning to the data as well i guess (so maybe that's why actually).

No I don't think so, my goal was just to ignore major updates while I implement the minor ones. But as @murdo-moj pointed out, I don't think there are any major updates possible when updating metadata, so actually it's more that we need to distinguish valid and invalid updates (for example changing the name or one of the generated fields).

> 2. Even with minor updates we'll still need to write all associated schema for the data product into the new version folder which isn't currently happening. I wrote this function to do that https://github.com/ministryofjustice/data-platform/blob/main/containers/daap-create-schema/src/var/task/create_schema.py#L18 for schema

Ah thanks. I was planning to implement the actual deletion in the lambda, so that this class only manages the metadata files. Does that make sense to you?

LavMatt commented:
> Ah thanks. I was planning to implement the actual deletion in the lambda, so that this class only manages the metadata files. Does that make sense to you?

Should there be any deletions? Yeah that does make sense to have the wider collection of metadata and schema managed outside of the class. Suppose we could have a function that copies everything from the previous version folder except files (either data product metadata or table schema) that already exist in the new versions folder. Could live in the paths module possibly

MatMoore commented:
> > Ah thanks. I was planning to implement the actual deletion in the lambda, so that this class only manages the metadata files. Does that make sense to you?
> 
> Should there be any deletions? Yeah that does make sense to have the wider collection of metadata and schema managed outside of the class. Suppose we could have a function that copies everything from the previous version folder except files (either data product metadata or table schema) that already exist in the new versions folder. Could live in the paths module possibly

Oops sorry I meant copying, not deletion! A function that copies everything might be useful, I think I'll have a go at implementing it in the lambda first and then if I have anything reusable I'll move it.

MatMoore committed:
Code review suggestion

Modify DataProductMetadata in place rather than creating a copy.

This is consistent with other methods on this class returning self.

MatMoore committed:
Fix type

MatMoore committed:
Code review suggestion: rename bucket_path -> latest_version_path

MatMoore committed:
Increment to new major version
